src/ripple/core/JobCounter.h

```diff
@@ -0,0 +1,188 @@
+// allows a Stoppable to assure that all child Jobs in the JobQueue are
+// completed before the Stoppable declares itself stopped().
+class JobCounter
+{
+private:
+    std::mutex mutable mutex_ {};
+    std::condition_variable allJobsDoneCond_ {};  // guard with mutex_
+    bool waitForJobs_ {false};                    // guard with mutex_
+    int jobCount_ {0};                            // guard with mutex_
+
+    // Manage the count.  If we're stopping and the count drops to zero
+    // notify allJobsDoneCond_.
+    void adjustCount (int upDn)
+    {
+        std::lock_guard<std::mutex> lock {mutex_};
+
+        // Count the job.  Notify if we're stopping and all jobs are done.
+        jobCount_ += upDn > 0 ? 1 : -1;
+        if (waitForJobs_ && jobCount_ == 0)
+            allJobsDoneCond_.notify_all();
+    }
+
+    // A private template class that helps count the number of Jobs
+    // in flight.  This allows Stoppables to hold off declaring stopped()
+    // until all their JobQueue Jobs are dispatched.
+    template <typename JobHandler>
+    class CountedJob
+    {
+    private:
+        JobCounter& jobCounter_;
+        JobHandler jobHandler_;
+
+        // A helper for the following static_assert.
+        template <typename T>
+        static typename std::remove_reference<T>::type&
+        toLvalueRef (T&& arg)
+        {
+            using ReturnType = typename std::remove_reference<T>::type&;
+            return static_cast<ReturnType>(arg);
+        }
+
+        static_assert (
+            std::is_same<decltype(
+                jobHandler_(toLvalueRef(Job{}))), void>::value,
+                "JobHandler must be callable with Job&");
+
+    public:
+        CountedJob () = delete;
+
+        CountedJob (CountedJob const& rhs)
+        : jobCounter_ (rhs.jobCounter_)
+        , jobHandler_ (rhs.jobHandler_)
+        {
+            jobCounter_.adjustCount (1);
+        }
+
+        CountedJob (CountedJob&& rhs)
+        : jobCounter_ (rhs.jobCounter_)
+        , jobHandler_ (std::move (rhs.jobHandler_))
+        {
+            jobCounter_.adjustCount (1);
+        }
+
+        CountedJob (JobCounter& jobCounter, JobHandler&& jobHandler)
+        : jobCounter_ (jobCounter)
+        , jobHandler_ (std::move (jobHandler))
+        {
+            jobCounter_.adjustCount (1);
+        }
+
+        CountedJob& operator=(CountedJob const& rhs) = delete;
+        CountedJob& operator=(CountedJob&& rhs) = delete;
+
+        ~CountedJob ()
+        {
+            jobCounter_.adjustCount (-1);
+        }
+
+        void operator ()(Job& job)
+        {
+            return jobHandler_ (job);
+        }
+    };
+
+public:
+    JobCounter() = default;
+    // Not copyable because outstanding counts would be hard to sort out.
+    JobCounter (JobCounter const&) = delete;
+    // Not movable because in-flight counted Jobs have a pointer to this.
+    JobCounter (JobCounter&&) = delete;
```

Comment:
This is a stylistic comment, but I'm on record as advising to never delete the move members.  Here is the rationale for that guideline: http://stackoverflow.com/a/38820178/576911  This is what that guideline calls the redundant case.  This isn't incorrect.  One just has to find the deleted copy constructor (which is very nicely placed in this code) to read that this is a correct use of deleted move members.

This isn't something I feel strongly about.  I just felt obliged to comment on this since I've gone and made a big public deal out of it. :-)